File: 프로그래머스/2/138476. 귤 고르기/귤 고르기.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

int solution(int k, vector<int> tangerine) {
    int answer = 0;
    
    // 크기:개수 o(n)
    unordered_map<int, int> size_cnt;
    for(int i = 0; i < tangerine.size(); i++){
        size_cnt[tangerine[i]]++;
    }
    
    // 개수 내림차순 o(nlogn)
    vector<int> nums;
    for(auto& m: size_cnt){
        nums.push_back(m.second);
    }
    
    // 내림차순 정렬
    sort(nums.begin(), nums.end(), greater<int>());
    
    for(int i = 0; i < nums.size(); i++){
        if(k <= 0) break;
        else{
            k-= nums[i];
            answer = i+1;
        }
    }
    
    
    return answer; // k개를 고를 때 크기가 서로 다른 종류의 수의 최솟값
}

File: 프로그래머스/2/138476. 귤 고르기/귤 고르기.java
import java.util.*;

class Solution {
    
    // k개 골라 상자에 담아 판매. 상자에 서로 다른 종류의 수를 최소화
    public int solution(int k, int[] tangerine) {
        int answer = 0;
        
        // {크기, 개수}
        HashMap<Integer, Integer> nums = new HashMap<>();
        for(int t : tangerine){
            nums.put(t, nums.getOrDefault(t, 0)+1);
        }
        List<Integer> sizes = new ArrayList<>(nums.values());
        sizes.sort((a, b) -> {
            return Integer.compare(b, a);
        }); 
        
        // 가장 크기가 큰 것부터 개수 세기
        int kLeft = k; 
        for(int siz : sizes){
            answer++;
            kLeft -= siz; 
            if(kLeft <= 0) break;
        }
        
        
        // 귤 k개를 고를 때 크기가 서로 다른 종류의 수의 최솟값을 return
        return answer;
    }
}